From 084fde2300d21a19ad7f6127f5f3b41db4a2b154 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 12 Mar 2024 18:06:04 +0100
Subject: [PATCH 0802/1695] drivers: regulator: use mutex_pm_aware

Use newly introduced struct mutex_pm_aware semaphore to protect
regulator accesses.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a6f12e2696e0c954e34a8d5d88d989886e58f568 drivers: regulator: use mutex_pm_aware]
Change-Id: I46cf6e32362604b3762e3b3e193ec4ac8e12a660
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
(cherry picked from commit c80790fe23c516a4ab7c7c640868539c12539093)
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/368045
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/regulator/regulator.c | 27 ++++++---------------------
 core/include/drivers/regulator.h   |  6 +++---
 2 files changed, 9 insertions(+), 24 deletions(-)

diff --git a/core/drivers/regulator/regulator.c b/core/drivers/regulator/regulator.c
index 67d3c98de..6ccedb699 100644
--- a/core/drivers/regulator/regulator.c
+++ b/core/drivers/regulator/regulator.c
@@ -11,7 +11,7 @@
 #include <keep.h>
 #include <kernel/boot.h>
 #include <kernel/delay.h>
-#include <kernel/mutex.h>
+#include <kernel/mutex_pm_aware.h>
 #include <kernel/panic.h>
 #include <kernel/pm.h>
 #include <kernel/tee_time.h>
@@ -26,32 +26,15 @@
 static SLIST_HEAD(, regulator) regulator_device_list =
 	SLIST_HEAD_INITIALIZER(regulator);
 
+/* Access protection mutex complying the power state transitions context */
 static void lock_regulator(struct regulator *regulator)
 {
-	/*
-	 * Regulator operation may occur at runtime and during specific
-	 * system power transition: power off, PM suspend and resume.
-	 * These operate upon fastcall entries, under PSCI services
-	 * execution, where non-secure world is not operational. In these
-	 * cases we cannot take a mutex and will expect the mutex is
-	 * unlocked.
-	 */
-	if (thread_get_id_may_fail() == THREAD_ID_INVALID) {
-		assert(!regulator->lock.state);
-		return;
-	}
-
-	mutex_lock(&regulator->lock);
+	mutex_pm_aware_lock(&regulator->mutex);
 }
 
 static void unlock_regulator(struct regulator *regulator)
 {
-	if (thread_get_id_may_fail() == THREAD_ID_INVALID) {
-		/* Path for PM sequences when with local Monitor */
-		return;
-	}
-
-	mutex_unlock(&regulator->lock);
+	mutex_pm_aware_unlock(&regulator->mutex);
 }
 
 static TEE_Result set_state(struct regulator *regulator, bool on_not_off)
@@ -288,6 +271,8 @@ TEE_Result regulator_register(struct regulator *regulator)
 	    regulator->flags & ~REGULATOR_FLAGS_MASK)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	mutex_pm_aware_init(&regulator->mutex);
+
 	regulator_get_range(regulator, &min_uv, &max_uv);
 	if (min_uv > max_uv)
 		return TEE_ERROR_BAD_PARAMETERS;
diff --git a/core/include/drivers/regulator.h b/core/include/drivers/regulator.h
index 5bfce0899..cb4a2ad26 100644
--- a/core/include/drivers/regulator.h
+++ b/core/include/drivers/regulator.h
@@ -7,7 +7,7 @@
 
 #include <assert.h>
 #include <bitstring.h>
-#include <kernel/mutex.h>
+#include <kernel/mutex_pm_aware.h>
 #include <sys/queue.h>
 #include <tee_api_types.h>
 #include <stdbool.h>
@@ -97,7 +97,7 @@ struct regulator_voltages_desc {
  * @enable_ramp_delay_us: Delay after enable, in microseconds (us)
  * @flags: REGULATOR_* property flags
  * @refcount: Regulator enable request reference counter
- * @lock: Mutex for concurrent access protection
+ * @mutex: Concurrent access protection considering PM context sequences
  * @voltages_fallback: Default supported voltage range description
  * @link: Link in initialized regulator list
  */
@@ -114,7 +114,7 @@ struct regulator {
 	/* Fields internal to regulator framework */
 	unsigned int flags;
 	unsigned int refcount;
-	struct mutex lock;	/* Concurrent access protection */
+	struct mutex_pm_aware mutex;
 	struct voltages_fallback {
 		struct regulator_voltages_desc desc;
 		int levels[3];
-- 
2.47.2

